add Half support for unary ops on CPU

jiayisunx opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99002
* __->__ #98493



Add Half support for log_sigmoid and some unary ops on CPU, including sinc, acosh, asinh, atanh, digamma, trigamma, rsqrt, acos, asin, atan, ceil, cos, erf, erfc, erfinv, exp, expml, floor, log, log10, log1p, log2, i0, round, sin, sqrt, tan, tanh, trunc, lgamma.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @peterbell10 @desertfire

jiayisunx committed:
add Half support for sinc, acosh, asinh, atanh, digamma and trigamma on CPU

[ghstack-poisoned]

jiayisunx committed:
Update on "add Half support for sinc, acosh, asinh, atanh, digamma and trigamma on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jiayisunx commented:
@pytorchbot label "topic: not user facing"

jiayisunx committed:
Update on "add Half support for sinc, acosh, asinh, atanh, digamma and trigamma on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jiayisunx committed:
Update on "add Half support for sinc, acosh, asinh, atanh, digamma and trigamma on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jiayisunx committed:
Update on "add Half support for sinc, acosh, asinh, atanh, digamma and trigamma on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jiayisunx committed:
Update on "add Half support for sinc, acosh, asinh, atanh, digamma and trigamma on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jiayisunx committed:
Update on "add Half support for sinc, acosh, asinh, atanh, digamma and trigamma on CPU"




cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jgong5 commented on c10/util/Half-math.h:
I'm wondering if we can share the implementation with bfloat16 counterpart considering they share the same implementation except for data types.

jiayisunx replied:
Sharing the same implementation is possible, but c10/util/Half-inl.h and c10/util/Half.cpp are implemented separately on two data types, would it be weird to just combine c10/util/Half-math.h and c10/util/BFloat16-math.h into one file?

jgong5 replied:
> would it be weird to just combine c10/util/Half-math.h and c10/util/BFloat16-math.h into one file?

It depends on how similar the logic looks like. Currently, the math part is almost the same. It should be ok to dedup them.

jiayisunx replied:
OK, thanks.

jiayisunx committed:
Update on "add Half support for unary ops on CPU"




Add Half support for log_sigmoid and some unary ops on CPU, including sinc, acosh, asinh, atanh, digamma, trigamma, rsqrt, acos, asin, atan, ceil, cos, erf, erfc, erfinv, exp, expml, floor, log, log10, log1p, log2, i0, round, sin, sqrt, tan, tanh, trunc, lgamma.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv peterbell10 desertfire

[ghstack-poisoned]

jiayisunx committed:
Update on "add Half support for unary ops on CPU"




Add Half support for log_sigmoid and some unary ops on CPU, including sinc, acosh, asinh, atanh, digamma, trigamma, rsqrt, acos, asin, atan, ceil, cos, erf, erfc, erfinv, exp, expml, floor, log, log10, log1p, log2, i0, round, sin, sqrt, tan, tanh, trunc, lgamma.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv peterbell10 desertfire

[ghstack-poisoned]

jiayisunx commented:
@pytorchbot merge

mingfeima commented on c10/util/BFloat16-math.h:
```diff
@@ -10,95 +11,188 @@ C10_CLANG_DIAGNOSTIC_IGNORE("-Wimplicit-float-conversion")
-/// Used by vec256<c10::BFloat16>::map
-inline c10::BFloat16 acos(c10::BFloat16 a) {
+template <typename T>
+struct is_reduced_floating_point
```
This has been defined in `ATen/cpu/vec/vec_base.h`, you may just include `<ATen/cpu/vec/vec.h>` to use it.

mingfeima commented on c10/util/BFloat16-math.h:
```diff
@@ -10,95 +11,188 @@ C10_CLANG_DIAGNOSTIC_IGNORE("-Wimplicit-float-conversion")
+
+template <
+    typename T,
+    typename std::enable_if_t<is_reduced_floating_point<T>::value, int> = 0>
```
nit: use `is_reduced_floating_point_v` here and the rest of code

jiayisunx replied:
Done, thanks.

mingfeima reviewed (changes requested):
Generally LGTM.

Just not sure whether we should overload `c10/util/BFloat16-math.h` or make a copy, say `c10/util/Half-math.h` ? @jgong5 ideas ?

jgong5 commented:
> Just not sure whether we should overload `c10/util/BFloat16-math.h` or make a copy, say `c10/util/Half-math.h` ? @jgong5 ideas ?

Do you mean rename the header file with a more general name? The implementation of these math functions is mostly identical for bfloat16 and half so we want to share code here.

EikanWang commented on test/inductor/test_torchinductor_opinfo.py:
```diff
@@ -186,11 +186,11 @@ def process(device_type):
     "nn.functional.local_response_norm": {i64},
     "nn.functional.one_hot": {i64},
     "nn.functional.rrelu": {f32, f64},
-    "nn.functional.triplet_margin_with_distance_loss": {f32, f64, i32, i64},
```
Regarding this operation and `gaussian_nll_loss`, the reason why we add `fp16` is that the two operations support FP16 now.

jiayisunx committed:
Update on "add Half support for unary ops on CPU"




Add Half support for log_sigmoid and some unary ops on CPU, including sinc, acosh, asinh, atanh, digamma, trigamma, rsqrt, acos, asin, atan, ceil, cos, erf, erfc, erfinv, exp, expml, floor, log, log10, log1p, log2, i0, round, sin, sqrt, tan, tanh, trunc, lgamma.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv peterbell10 desertfire

[ghstack-poisoned]

jiayisunx committed:
Update on "add Half support for unary ops on CPU"




Add Half support for log_sigmoid and some unary ops on CPU, including sinc, acosh, asinh, atanh, digamma, trigamma, rsqrt, acos, asin, atan, ceil, cos, erf, erfc, erfinv, exp, expml, floor, log, log10, log1p, log2, i0, round, sin, sqrt, tan, tanh, trunc, lgamma.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv peterbell10 desertfire

[ghstack-poisoned]

jiayisunx committed:
Update on "add Half support for unary ops on CPU"




Add Half support for log_sigmoid and some unary ops on CPU, including sinc, acosh, asinh, atanh, digamma, trigamma, rsqrt, acos, asin, atan, ceil, cos, erf, erfc, erfinv, exp, expml, floor, log, log10, log1p, log2, i0, round, sin, sqrt, tan, tanh, trunc, lgamma.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv peterbell10 desertfire

[ghstack-poisoned]

jiayisunx committed:
Update on "add Half support for unary ops on CPU"




Add Half support for log_sigmoid and some unary ops on CPU, including sinc, acosh, asinh, atanh, digamma, trigamma, rsqrt, acos, asin, atan, ceil, cos, erf, erfc, erfinv, exp, expml, floor, log, log10, log1p, log2, i0, round, sin, sqrt, tan, tanh, trunc, lgamma.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv peterbell10 desertfire

[ghstack-poisoned]

jiayisunx committed:
Update on "add Half support for unary ops on CPU"




Add Half support for log_sigmoid and some unary ops on CPU, including sinc, acosh, asinh, atanh, digamma, trigamma, rsqrt, acos, asin, atan, ceil, cos, erf, erfc, erfinv, exp, expml, floor, log, log10, log1p, log2, i0, round, sin, sqrt, tan, tanh, trunc, lgamma.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv peterbell10 desertfire

[ghstack-poisoned]

jiayisunx committed:
Update on "add Half support for unary ops on CPU"




Add Half support for log_sigmoid and some unary ops on CPU, including sinc, acosh, asinh, atanh, digamma, trigamma, rsqrt, acos, asin, atan, ceil, cos, erf, erfc, erfinv, exp, expml, floor, log, log10, log1p, log2, i0, round, sin, sqrt, tan, tanh, trunc, lgamma.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv peterbell10 desertfire

[ghstack-poisoned]

jiayisunx committed:
Update on "add Half support for unary ops on CPU"




Add Half support for log_sigmoid and some unary ops on CPU, including sinc, acosh, asinh, atanh, digamma, trigamma, rsqrt, acos, asin, atan, ceil, cos, erf, erfc, erfinv, exp, expml, floor, log, log10, log1p, log2, i0, round, sin, sqrt, tan, tanh, trunc, lgamma.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv peterbell10 desertfire

[ghstack-poisoned]

jiayisunx committed:
Update on "add Half support for unary ops on CPU"




Add Half support for log_sigmoid and some unary ops on CPU, including sinc, acosh, asinh, atanh, digamma, trigamma, rsqrt, acos, asin, atan, ceil, cos, erf, erfc, erfinv, exp, expml, floor, log, log10, log1p, log2, i0, round, sin, sqrt, tan, tanh, trunc, lgamma.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv peterbell10 desertfire

[ghstack-poisoned]

jiayisunx committed:
Update on "add Half support for unary ops on CPU"




Add Half support for log_sigmoid and some unary ops on CPU, including sinc, acosh, asinh, atanh, digamma, trigamma, rsqrt, acos, asin, atan, ceil, cos, erf, erfc, erfinv, exp, expml, floor, log, log10, log1p, log2, i0, round, sin, sqrt, tan, tanh, trunc, lgamma.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv peterbell10 desertfire

[ghstack-poisoned]

jiayisunx committed:
Update on "add Half support for unary ops on CPU"




Add Half support for log_sigmoid and some unary ops on CPU, including sinc, acosh, asinh, atanh, digamma, trigamma, rsqrt, acos, asin, atan, ceil, cos, erf, erfc, erfinv, exp, expml, floor, log, log10, log1p, log2, i0, round, sin, sqrt, tan, tanh, trunc, lgamma.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv peterbell10 desertfire

[ghstack-poisoned]

jiayisunx committed:
Update on "add Half support for unary ops on CPU"




Add Half support for log_sigmoid and some unary ops on CPU, including sinc, acosh, asinh, atanh, digamma, trigamma, rsqrt, acos, asin, atan, ceil, cos, erf, erfc, erfinv, exp, expml, floor, log, log10, log1p, log2, i0, round, sin, sqrt, tan, tanh, trunc, lgamma.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv peterbell10 desertfire

[ghstack-poisoned]

jiayisunx committed:
Update on "add Half support for unary ops on CPU"




Add Half support for log_sigmoid and some unary ops on CPU, including sinc, acosh, asinh, atanh, digamma, trigamma, rsqrt, acos, asin, atan, ceil, cos, erf, erfc, erfinv, exp, expml, floor, log, log10, log1p, log2, i0, round, sin, sqrt, tan, tanh, trunc, lgamma.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv peterbell10 desertfire

[ghstack-poisoned]

jiayisunx committed:
Update on "add Half support for unary ops on CPU"




Add Half support for log_sigmoid and some unary ops on CPU, including sinc, acosh, asinh, atanh, digamma, trigamma, rsqrt, acos, asin, atan, ceil, cos, erf, erfc, erfinv, exp, expml, floor, log, log10, log1p, log2, i0, round, sin, sqrt, tan, tanh, trunc, lgamma.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv peterbell10 desertfire

[ghstack-poisoned]

jiayisunx committed:
Update on "add Half support for unary ops on CPU"




Add Half support for log_sigmoid and some unary ops on CPU, including sinc, acosh, asinh, atanh, digamma, trigamma, rsqrt, acos, asin, atan, ceil, cos, erf, erfc, erfinv, exp, expml, floor, log, log10, log1p, log2, i0, round, sin, sqrt, tan, tanh, trunc, lgamma.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv peterbell10 desertfire

[ghstack-poisoned]

jiayisunx committed:
Update on "add Half support for unary ops on CPU"




Add Half support for log_sigmoid and some unary ops on CPU, including sinc, acosh, asinh, atanh, digamma, trigamma, rsqrt, acos, asin, atan, ceil, cos, erf, erfc, erfinv, exp, expml, floor, log, log10, log1p, log2, i0, round, sin, sqrt, tan, tanh, trunc, lgamma.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv peterbell10 desertfire

[ghstack-poisoned]

jiayisunx committed:
Update on "add Half support for unary ops on CPU"




Add Half support for log_sigmoid and some unary ops on CPU, including sinc, acosh, asinh, atanh, digamma, trigamma, rsqrt, acos, asin, atan, ceil, cos, erf, erfc, erfinv, exp, expml, floor, log, log10, log1p, log2, i0, round, sin, sqrt, tan, tanh, trunc, lgamma.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv peterbell10 desertfire

[ghstack-poisoned]

jiayisunx committed:
Update on "add Half support for unary ops on CPU"




Add Half support for log_sigmoid and some unary ops on CPU, including sinc, acosh, asinh, atanh, digamma, trigamma, rsqrt, acos, asin, atan, ceil, cos, erf, erfc, erfinv, exp, expml, floor, log, log10, log1p, log2, i0, round, sin, sqrt, tan, tanh, trunc, lgamma.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv peterbell10 desertfire

[ghstack-poisoned]

jiayisunx committed:
Update on "add Half support for unary ops on CPU"




Add Half support for log_sigmoid and some unary ops on CPU, including sinc, acosh, asinh, atanh, digamma, trigamma, rsqrt, acos, asin, atan, ceil, cos, erf, erfc, erfinv, exp, expml, floor, log, log10, log1p, log2, i0, round, sin, sqrt, tan, tanh, trunc, lgamma.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv peterbell10 desertfire

[ghstack-poisoned]

jiayisunx committed:
Update on "add Half support for unary ops on CPU"




Add Half support for log_sigmoid and some unary ops on CPU, including sinc, acosh, asinh, atanh, digamma, trigamma, rsqrt, acos, asin, atan, ceil, cos, erf, erfc, erfinv, exp, expml, floor, log, log10, log1p, log2, i0, round, sin, sqrt, tan, tanh, trunc, lgamma.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv peterbell10 desertfire

[ghstack-poisoned]

jiayisunx committed:
Update on "add Half support for unary ops on CPU"




Add Half support for log_sigmoid and some unary ops on CPU, including sinc, acosh, asinh, atanh, digamma, trigamma, rsqrt, acos, asin, atan, ceil, cos, erf, erfc, erfinv, exp, expml, floor, log, log10, log1p, log2, i0, round, sin, sqrt, tan, tanh, trunc, lgamma.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv peterbell10 desertfire

[ghstack-poisoned]

jiayisunx committed:
Update on "add Half support for unary ops on CPU"




Add Half support for log_sigmoid and some unary ops on CPU, including sinc, acosh, asinh, atanh, digamma, trigamma, rsqrt, acos, asin, atan, ceil, cos, erf, erfc, erfinv, exp, expml, floor, log, log10, log1p, log2, i0, round, sin, sqrt, tan, tanh, trunc, lgamma.

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 soumith voznesenskym penguinwu anijain2305 EikanWang Guobing-Chen zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv peterbell10 desertfire

[ghstack-poisoned]

jiayisunx commented:
I found that torch.round float16 calculation are incorrect in mps backend, and I have to add 'round' into expected fail list as workarround, and I have reported a bug issue to pytorch: https://github.com/pytorch/pytorch/issues/101167

jiayisunx commented:
@pytorchbot merge

jiayisunx commented:
@ngimel, can you please review this PR? thanks!

jiayisunx commented:
@pytorchbot merge

Pitenager commented:
Hi team,

Do you plan to publish a new release soon?

I'm developing my MSc thesis and I'm facing this same issue `RuntimeError: "clamp_scalar_cpu" not implemented for 'Half'`

If not, is there any workaround? Downgrading is not a good one because I have other dependencies that require PyTorch >= 2.0.0

Thanks in advance
